### ISSUE TITLE
GCM Core on Windows cannot handle tokens larger than 1KB in size

### ISSUE THREAD

**Issue by AtOMiCNebula** [microsoft/Git-Credential-Manager-Core#190]
I've been trying to automate an Azure Pipeline to check out a repo with Scalar (v20.07.174.1) and GCMCore (v2.0.252-beta), and ran into a pretty weird issue.  The ephemeral access token that Azure Pipelines generates for the run usually squeak in just under 1KB in size (anecdotally, ~1,015 bytes).  However, when a pipeline uses a relatively new (to me, anyways) [cache task](https://docs.microsoft.com/en-us/azure/devops/pipelines/release/caching?view=azure-devops), the token grows to be larger than 1KB (anecdotally, ~1,420 bytes).  This causes `git credential approve` to start failing:

```
+ git credential approve
fatal: Failed to write item to store. [0x6f7]
fatal: The stub received bad data
```

"Failed to write item to store" comes from here: https://github.com/microsoft/Git-Credential-Manager-Core/blob/b3607300896f9a20e2bed1582216a33fb2ac8d13/src/shared/Microsoft.Git.CredentialManager/Interop/Windows/WindowsCredentialManager.cs#L92

Looking at nearby code, my guess is that the Windows credential store simply doesn't want to handle credentials that large.  Interestingly, the [CredWriteW documentation](https://docs.microsoft.com/en-us/windows/win32/api/wincred/ns-wincred-credentiala) I found suggests `CRED_MAX_CREDENTIAL_BLOB_SIZE` is 512 bytes, so maybe we're just getting lucky that it's *actually* letting us store twice that?

I suspect the solution here is for the Windows-specific code to handle spanning a token over multiple Windows credential entries.  I think I can work around this for now by using the `store` credential provider and just putting the token in a file for now...but it'd be better to have real support for large(r) tokens in GCMCore.

**Comment by AtOMiCNebula** [comment]
(I filed this as a "feature request" instead of an "authentication issue" because the latter went off into the weeds of "can _you_ access the repo" and this was more gritty than that...and there was no generic "bug" template to use, so here we are)

**Comment by kyle-rader** [comment]
(This issue discovered by Office)

**Comment by derrickstolee** [comment]
> The ephemeral access token that Azure Pipelines generates for the run usually squeak in just under 1KB in size (anecdotally, ~1,015 bytes). However, when a pipeline uses a relatively new (to me, anyways) [cache task](https://docs.microsoft.com/en-us/azure/devops/pipelines/release/caching?view=azure-devops), the token grows to be larger than 1KB (anecdotally, ~1,420 bytes).

Perhaps a dumb question: _why_ are these tokens so incredibly large? Wouldn't a token of 128 bytes be sufficiently large for secure purposes?

Maybe @vtbassmatt has some context as to why Pipelines generates such a large token.

**Comment by AtOMiCNebula** [comment]
JWTs are not small (base64-encoded small JSON payload + base64-encoded larger JSON payload + signature(s)), and the presence of the Cache task causes additional claims to be added to the JWT that pushes it past the 1024-byte boundary.

**Comment by derrickstolee** [comment]
Thanks for the clarity here that these are [JSON Web Tokens](https://jwt.io/), which I had never heard of before.

Just because what you are asking for is quite a significant change and is not something we expect _human users_ to need, I expect this to not be delivered very quickly.

To get you unblocked: is there a way you could use a different kind of token here?

**Comment by AtOMiCNebula** [comment]
I believe I won't be blocked if I switch to using the built-in `store` credential provider, instead of GCM/GCMCore, though I'll admit that I haven't had a chance to test that yet (though it has worked for me in the past!).  It's not the most elegant workaround, but it should get the job done. :smile:

The Azure Pipelines agent handles authenticating to Azure DevOps/GitHub by passing something like `-c http.extraheader="AUTHORIZATION: bearer ***"` to any network-utilizing Git command (which feels a little like cheating, but whatever).  Scalar/GVFS doesn't have that facility AFAIK, so I'm left with the credential provider route.

Using a different kind of token (PAT?) isn't a good option, since we have a lot of pipelines and it'll be impossible to keep the PAT secure.

**Comment by derrickstolee** [comment]
> The Azure Pipelines agent handles authenticating to Azure DevOps/GitHub by passing something like `-c http.extraheader="AUTHORIZATION: bearer ***"` to any network-utilizing Git command (which feels a little like cheating, but whatever). Scalar/GVFS doesn't have that facility AFAIK, so I'm left with the credential provider route.

You mean, we don't have the ability to pass this during a `scalar clone` command? I think you could fake it by running `git config --global http.extraheader "AUTHORIZATION: bearer ***"` before `scalar clone` (not tested).

My gut reaction is that we are more likely to support the `-c` syntax in Scalar before we split tokens in GCM Core across multiple stores.

**Comment by derrickstolee** [comment]
@AtOMiCNebula: this is more complicated than I'd like for shipping in the pending release of Scalar. Please see [this comment about `GIT_CONFIG_PARAMETERS`](https://github.com/microsoft/scalar/issues/452#issuecomment-712821461) as another workaround.

**Comment by AtOMiCNebula** [comment]
If you think this scenario is out of scope for GCMCore, then I'm fine with just falling back to `store`.  Not all of the agents themselves are ephemeral, but for those that are not, I can make sure the tokens don't get left behind (even if the service invalidates them anyways), which is needed regardless of us using GCMCore or some other solution. :smile:

**Comment by peter-esik** [comment]
I just wanted to throw in my two cents here: you don't need to do anything fancy or extreme to get tokens this big. A simple barebones Entra ID application's id token is `~2.6 KiB`, which is over the limit by a hair (according to the [docs](https://learn.microsoft.com/en-us/windows/win32/api/wincred/ns-wincred-credentialw) `CRED_MAX_CREDENTIAL_BLOB_SIZE` is `5*512`, so `2.5 KiB`).

After some guesswork, I finally searched for the pretty generic error message `Failed to write item to store. [0x6f7]`, and got to this issue.

A check if the credential to be stored is `> CRED_MAX_CREDENTIAL_BLOB_SIZE` and providing a more specific error message in this case would be nice.
